### DISCUSSION TITLE
Is my understanding correct that CLOffset represents the word offset? @rahulbera

### DISCUSSION THREAD

**Discussion by szuhzw** [CMU-SAFARI/Hermes#34] [Q&A] [answered]
![image](https://github.com/user-attachments/assets/eaaf7e12-510e-4222-ab33-e761f4bcea85)
![image](https://github.com/user-attachments/assets/549a5d2a-bef7-4b6d-9b2b-1a4c3be5d706)
![image](https://github.com/user-attachments/assets/d49d04af-3442-4b60-b660-db47f82775fb)
![image](https://github.com/user-attachments/assets/39282f1b-bffc-423c-99be-abe4d86ce50e)

**Comment by rahulbera** [comment, marked as answer]
HI @szuhzw. You are correct. This is a bug. Technically, `generate_index_from_feature` function should call `process_CLOffset` (or `process_PC_CLOffset`) for `CLOffset` (or `PC_CLOffset`) feature type. But it was mistakenly calling the `process_CLWordOffset`. The same bug is also present for `CLDWordOffset` feature too. This bug likely got in during refactoing the code for GitHub release. 

I am creating a patch to fix this issue. Thanks for pointing it out.

**Comment by rahulbera** [comment]
Hi @szuhzw, I merged this [pull request](https://github.com/CMU-SAFARI/Hermes/pull/36), which should fix this issue.

Now that this bug has been caught and fixed, can you please check and report what's the performance difference (in terms of accuracy, coverage, and overall IPC) of Hermes across PC_CLOffset, PC_CLWordOffset, PC_CLDWordOffset features? Thanks.

**Reply by szuhzw** [reply]
Hi @rahulbera , The latest perc_pred_helper.cc doesn't define the process_PC_CLDWordOffset function.

**Reply by szuhzw** [reply]
Hi @rahulbera , 
uint32_t process_PC_CLDwordOffset(state_info_t *state, uint64_t metadata, int32_t hash_type, uint32_t weight_array_size). 
Is it right that the process_PC_CLDwordOffset should be process_PC_CLDWordOffset?

**Reply by szuhzw** [reply]
Hi. @rahulbera , Do you mean you want me to test the performance of the following three sets of features? 
5,8,9,11,16
5,8,9,13,16
5,8,9,15,16

**Reply by rahulbera** [reply]
Hi @szuhzw, yes. There was a typo in the function name `process_PC_CLDwordOffset`. It's fixed now.

**Reply by rahulbera** [reply]
Regarding your second question: yes. Since the bug is fixed now, we should do a regression test with the three feature sets that you mentioned.

**Comment by szuhzw** [comment]
Hi. @rahulbera
I ran over 200 traces. In terms of accuracy, the difference between “5,8,9,11,16” and "5,8,9,13,16" is extremely small, while "5,8,9,15,16"'s accuracy is about 2% lower than “5,8,9,11,16” and "5,8,9,13,16". In terms of recall, the difference in coverage among these three sets of features is extremely small. In terms of single-core performance, the geometric mean IPC difference between these three feature sets is extremely small.

**Comment by rahulbera** [comment]
Thanks @szuhzw for running the regression. Good to know that the results are not differing much.

**Reply by szuhzw** [reply]
Hi. @rahulbera 
Has the research group you are in ever reviewed an paper titled "*****: An Ultra-Lightweight Level High-Performance Off-Chip Load Predictor for Long-Latency Load Requests"?
